### PR TITLE
Add a basic yield function to the scheduler

### PR DESCRIPTION
**Pull request by m-m-adams**
This is a very basic yield with no context switching. It allows recalling the yielding task to match how routine for sd is used in the sd read functions, so still needs guards (if (sdRoutineLock) )

### CODE REVIEW

File: src/OSLikeStuff/task_scheduler.cpp [hunk 1]
```diff
@@ -39,6 +43,13 @@ struct StatBlock {
 #endif
 		average = (average + v) / 2;
 	}
+	void reset() {
+#if SCHEDULER_DETAILED_STATS
+		min = std::numeric_limits<double>::infinity();
+		max = -std::numeric_limits<double>::infinity();
```
Comment (seangoodvibes) [inline src/OSLikeStuff/task_scheduler.cpp:49]: why is min positive and max negative?

File: src/deluge/drivers/sd/sd.c [hunk 2]
```diff
@@ -38,10 +38,11 @@ int32_t sddev_power_on(int32_t sd_port) {
...
 	sddev_start_timer(100); /* wait 100ms */
-	while (sddev_check_timer() == SD_OK) {
-		/* wait */
-		routineForSD(); // By Rohan
-	}
+	                        //	while (sddev_check_timer() == SD_OK) {
+	                        //		/* wait */
+	                        //		routineForSD(); // By Rohan
+	                        //	}
+	yieldingRoutineForSD((RunCondition)sddev_check_timer);
```
Comment (seangoodvibes) [inline src/deluge/drivers/sd/sd.c:45]: do we need to keep the old code?

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 3]
```diff
@@ -227,272 +227,234 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
-			AudioEngine::songSwapAboutToHappen();
-			AudioEngine::logAction("b");
-			playbackHandler.songSwapShouldPreserveTempo = Buttons::isButtonPressed(deluge::hid::button::TEMPO_ENC);
-		}
+		deleteOldSongBeforeLoadingNew();
+	}
+	else {
+		// Note: this is dodgy, but in this case we don't reset view.activeControllableClip here - we let the user keep
+		// fiddling with it. It won't get deleted.
+		AudioEngine::logAction("a");
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/load/load_song_ui.cpp:274]: I know this is old code, but logging "a" and "b" seems not very informative?

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 4]
```diff
@@ -227,272 +227,234 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
+		delugeDealloc(toDealloc);
+		preLoadedSong = NULL;
+		goto someError;
+	}
 
-		GlobalEffectable::initParams(&preLoadedSong->paramManager);
+	GlobalEffectable::initParams(&preLoadedSong->paramManager);
 
-		AudioEngine::logAction("c");
+	AudioEngine::logAction("c");
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/load/load_song_ui.cpp:318]: same comment above about log detail

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 5]
```diff
@@ -227,272 +227,234 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
-			goto gotErrorAfterCreatingSong;
-		}
-		AudioEngine::logAction("d");
+	// Will return false if we ran out of RAM. This isn't currently detected for while loading ParamNodes, but chances
+	// are, after failing on one of those, it'd try to load something else and that would fail.
+	error = preLoadedSong->readFromFile(smDeserializer);
+	if (error != Error::NONE) {
+		goto gotErrorAfterCreatingSong;
+	}
+	AudioEngine::logAction("d");
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/load/load_song_ui.cpp:327]: same here - log detail

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 6]
```diff
@@ -227,272 +227,234 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
+	if (playbackHandler.isEitherClockActive()) {
+		preLoadedSong->loadCrucialSamplesOnly();
+	}
+	else {
+		preLoadedSong->loadAllSamples(true);
+	}
 
-		preLoadedSong->name.set(&enteredText);
+	// Ensure all AudioFile Clusters needed for new song are loaded
+	static int32_t count =
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/load/load_song_ui.cpp:364]: opportunity to fix comment wrapping :D?

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 7]
```diff
@@ -227,272 +227,234 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
+			else {
+				display->setText("DONE", false, 255, true, NULL, false, true);
+			}
+			currentUIMode = UI_MODE_LOADING_SONG_UNESSENTIAL_SAMPLES_UNARMED;
 		}
+
+		// We're now waiting, either for the user to arm, or for the arming to launch the song-swap. Get loading all the
+		// rest of the samples which weren't needed right away. (though we might run out of RAM cos we haven't discarded
+		// all the old samples yet)
+		AudioEngine::logAction("g");
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/load/load_song_ui.cpp:411]: more log's..maybe it's fine?

File: src/deluge/gui/ui/load/load_song_ui.cpp [hunk 8]
```diff
@@ -227,272 +227,234 @@ ActionResult LoadSongUI::buttonAction(deluge::hid::Button b, bool on, bool inCar
...
-		// Delete the old song
-		AudioEngine::logAction("i");
-		if (toDelete) {
-			void* toDealloc = dynamic_cast<void*>(toDelete);
-			toDelete->~Song();
-			delugeDealloc(toDealloc);
-		}
+	if (display->haveOLED()) {
+		deluge::hid::display::OLED::displayWorkingAnimation(
+		    "Loading"); // To override our popup if we did one. (Still necessary?)
```
Comment (seangoodvibes) [inline src/deluge/gui/ui/load/load_song_ui.cpp:430]: comment wrapping :D

File: src/OSLikeStuff/task_scheduler.cpp (hunk 1 above)
Comment (m-m-adams) [reply src/OSLikeStuff/task_scheduler.cpp:49]: They're initial values - for min the algorithm is going to always make this value smaller so it needs to start by overestimating

File: src/deluge/drivers/sd/sd.c (hunk 2 above)
Comment (m-m-adams) [reply src/deluge/drivers/sd/sd.c:45]: Yeah it's to make it easier to go back if we need to remove the scheduler for the 1.2 release

Comment (m-m-adams) [reply src/deluge/drivers/sd/sd.c:45]: This change isn't supposed to be in this PR, it belongs in the other one. Pulling it out

**Comment by m-m-adams** [comment]
Force push was to remove the sd reading change - that belongs in #2111

**Review by seangoodvibes — APPROVED** [review]

**Pull request merged**